tokio-util/src/sync/cancellation_token.rs

```diff
@@ -318,561 +193,311 @@ impl<'a> core::fmt::Debug for WaitForCancellationFuture<'a> {
+    ///
+    /// Returns the (now parentless) children.
+    fn disconnect_children(node: &mut Inner) -> Vec<Arc<TreeNode>> {
+        for child in &node.children {
+            let mut child = child.inner.lock().unwrap();
+            child.parent_idx = 0;
+            child.parent = None;
         }
-    }
-
-    /// Whether this `CancellationTokenState` is still referenced by any
-    /// `CancellationToken`.
-    fn has_refs(&self) -> bool {
-        self.refcount != 0 || self.has_parent_ref
-    }
-}
 
-/// The maximum permitted amount of references to a CancellationToken. This
-/// is derived from the intent to never use more than 32bit in the `Snapshot`.
-const MAX_REFS: u32 = (std::u32::MAX - 7) >> 3;
-
-/// Internal state of the `CancellationToken` pair above
-struct CancellationTokenState {
-    state: AtomicUsize,
-    parent: Option<NonNull<CancellationTokenState>>,
-    from_parent: SynchronizedThroughParent,
-    synchronized: Mutex<SynchronizedState>,
-}
-
-impl CancellationTokenState {
-    fn new(
-        parent: Option<NonNull<CancellationTokenState>>,
-        state: StateSnapshot,
-    ) -> CancellationTokenState {
-        CancellationTokenState {
-            parent,
-            from_parent: SynchronizedThroughParent {
-                prev_peer: None,
-                next_peer: None,
-            },
-            state: AtomicUsize::new(state.pack()),
-            synchronized: Mutex::new(SynchronizedState::new()),
-        }
-    }
+        let mut dropped_children = vec![];
+        std::mem::swap(&mut node.children, &mut dropped_children);
 
-    /// Returns a snapshot of the current atomic state of the token
-    fn snapshot(&self) -> StateSnapshot {
-        StateSnapshot::unpack(self.state.load(Ordering::SeqCst))
+        dropped_children
     }
 
-    fn atomic_update_state<F>(&self, mut current_state: StateSnapshot, func: F) -> StateSnapshot
-    where
-        F: Fn(StateSnapshot) -> StateSnapshot,
-    {
-        let mut current_packed_state = current_state.pack();
+    /// Figures out the parent of the node and locks the node and its parent atomically.
+    ///
+    /// The basic principle of preventing deadlocks in the tree is
+    /// that we always lock the parent first, and then the child.
+    ///
+    /// Sadly, it's impossible to figure out the parent of a node without
+    /// locking it. To then achieve locking order consistency, the node
+    /// has to be unlocked before the parent gets locked.
+    /// This leaves a small window where we already assume that we know the parent,
+    /// but neither the parent or the node is locked. Therefore, the parent could change.
+    ///
+    /// To prevent that this problem leaks into the rest of the code, it is abstracted
+    /// in this function.
+    fn with_locked_node_and_parent<
+        F: FnOnce(MutexGuard<'_, Inner>, Option<MutexGuard<'_, Inner>>) -> Ret,
+        Ret,
+    >(
+        node: &Arc<TreeNode>,
+        func: F,
+    ) -> Ret {
```

Comment:
```suggestion
    fn with_locked_node_and_parent<F, Ret>(
        node: &Arc<TreeNode>,
        func: F,
    ) -> Ret
    where
        F: FnOnce(&mut Inner, Option<&mut Inner>) -> Ret,
    {
```

Reply:
The `MutexGuard` is in there on purpose, so that `func` can choose to unlock either if required. (See `remove_child`)

Reply:
Updated